mypy/plugins/default.py

```diff
@@ -107,55 +107,65 @@ def get_class_decorator_hook(self, fullname: str
 
 
 def open_callback(ctx: FunctionContext) -> Type:
-    """Infer a better return type for 'open'."""
-    return _analyze_open_signature(
+    """Verify argument types for 'open'."""
+    _verify_open_signature(
         arg_types=ctx.arg_types,
         args=ctx.args,
+        arg_names=ctx.callee_arg_names,
         mode_arg_index=1,
-        default_return_type=ctx.default_return_type,
+        text_only_arg_indices=(3, 4, 5),
         api=ctx.api,
+        context=ctx.context,
     )
+    return ctx.default_return_type
 
 
 def path_open_callback(ctx: MethodContext) -> Type:
-    """Infer a better return type for 'pathlib.Path.open'."""
-    return _analyze_open_signature(
+    """Verify argument types for 'pathlib.Path.open'."""
+    _verify_open_signature(
         arg_types=ctx.arg_types,
         args=ctx.args,
+        arg_names=ctx.callee_arg_names,
         mode_arg_index=0,
-        default_return_type=ctx.default_return_type,
+        text_only_arg_indices=(2, 3, 4),
         api=ctx.api,
+        context=ctx.context,
     )
+    return ctx.default_return_type
 
 
-def _analyze_open_signature(arg_types: List[List[Type]],
-                            args: List[List[Expression]],
-                            mode_arg_index: int,
-                            default_return_type: Type,
-                            api: CheckerPluginInterface,
-                            ) -> Type:
-    """A helper for analyzing any function that has approximately
-    the same signature as the builtin 'open(...)' function.
-
-    Currently, the only thing the caller can customize is the index
-    of the 'mode' argument. If the mode argument is omitted or is a
-    string literal, we refine the return type to either 'TextIO' or
-    'BinaryIO' as appropriate.
+def _verify_open_signature(
+        arg_types: List[List[Type]],
+        args: List[List[Expression]],
+        arg_names: List[Optional[str]],
+        mode_arg_index: int,
+        text_only_arg_indices: Tuple[int, ...],
+        api: CheckerPluginInterface,
+        context: Context,
+) -> None:
+    """A helper for verifying any function that has approximately the same
+    signature as the builtin 'open(...)' function.
+
+    If mode is detected to be binary, verify that text-only arguments
+    (encoding, errors, newline) are None.
     """
-    mode = None
     if not arg_types or len(arg_types[mode_arg_index]) != 1:
-        mode = 'r'
-    else:
-        mode_expr = args[mode_arg_index][0]
-        if isinstance(mode_expr, StrExpr):
-            mode = mode_expr.value
-    if mode is not None:
-        assert isinstance(default_return_type, Instance)  # type: ignore
-        if 'b' in mode:
-            return api.named_generic_type('typing.BinaryIO', [])
-        else:
-            return api.named_generic_type('typing.TextIO', [])
-    return default_return_type
+        return None
```

Comment:
Style nit: we prefer a bare `return` with no `None` when the return type is annotated as `None`.